File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.c
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// Helper to calculate multinomial coefficient capped at threshold
long long getPermutations(int* counts, int total, int threshold) {
    long long res = 1;
    int currentTotal = 1;
    
    for (int i = 0; i < 26; i++) {
        int cnt = counts[i];
        for (int j = 1; j <= cnt; j++) {
            res = res * currentTotal / j;
            currentTotal++;
            if (res > threshold) {
                return (long long)threshold + 1; // Cap to avoid integer overflow
            }
        }
    }
    return res;
}

char* smallestPalindrome(char* s, int k) {
    int n = strlen(s);
    int totalCounts[26] = {0};
    for (int i = 0; i < n; i++) {
        totalCounts[s[i] - 'a']++;
    }
    
    int halfLen = n / 2;
    int halfCounts[26] = {0};
    char midChar = 0;
    
    for (int i = 0; i < 26; i++) {
        halfCounts[i] = totalCounts[i] / 2;
        if (totalCounts[i] % 2 != 0) {
            midChar = (char)(i + 'a');
        }
    }
    
    char* firstHalf = (char*)malloc(halfLen * sizeof(char));
    int remainingSlots = halfLen;
    
    // Build the first half position by position
    for (int pos = 0; pos < halfLen; pos++) {
        bool found = false;
        for (int c = 0; c < 26; c++) {
            if (halfCounts[c] > 0) {
                halfCounts[c]--;
                remainingSlots--;
                
                long long p = getPermutations(halfCounts, remainingSlots, k);
                
                if (p >= k) {
                    firstHalf[pos] = (char)(c + 'a');
                    found = true;
                    break;
                } else {
                    k -= p;
                    halfCounts[c]++;
                    remainingSlots++;
                }
            }
        }
        if (!found) {
            free(firstHalf);
            return "";
        }
    }
    
    if (k > 1) {
        free(firstHalf);
        return "";
    }
    
    // Reconstruct full mirror buffer string
    char* result = (char*)malloc((n + 1) * sizeof(char));
    result[n] = '\0';
    
    for (int i = 0; i < halfLen; i++) {
        result[i] = firstHalf[i];
        result[n - 1 - i] = firstHalf[i];
    }
    if (n % 2 != 0) {
        result[halfLen] = midChar;
    }
    
    free(firstHalf);
    return result;
}

File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.py
class Solution:
    def smallestPalindrome(self, s: str, k: int) -> str:
        n = len(s)
        total_counts = [0] * 26
        for char in s:
            total_counts[ord(char) - 97] += 1
            
        half_len = n // 2
        half_counts = [0] * 26
        mid_char = ""
        
        for i in range(26):
            half_counts[i] = total_counts[i] // 2
            if total_counts[i] % 2 != 0:
                mid_char = chr(i + 97)
                
        first_half = [""] * half_len
        remaining_slots = half_len
        
        # Build the first half position by position
        for pos in range(half_len):
            found = False
            for c in range(26):
                if half_counts[c] > 0:
                    half_counts[c] -= 1
                    remaining_slots -= 1
                    
                    # Calculate remaining permutations capping at k + 1 to avoid overflow
                    p = self.getPermutations(half_counts, remaining_slots, k)
                    
                    if p >= k:
                        first_half[pos] = chr(c + 97)
                        found = True
                        break # Character locked into place
                    else:
                        k -= p
                        half_counts[c] += 1
                        remaining_slots += 1
            if not found:
                return "" # Fewer than k permutations available
                
        if k > 1:
            return ""
            
        # Reconstruct mirrored string via fast python slices
        half_str = "".join(first_half)
        return half_str + mid_char + half_str[::-1]

    # Helper to calculate multinomial coefficient capped at threshold
    def getPermutations(self, counts: list, total: int, threshold: int) -> int:
        res = 1
        current_total = 1
        
        for i in range(26):
            cnt = counts[i]
            for j in range(1, cnt + 1):
                res = (res * current_total) // j
                current_total += 1
                if res > threshold:
                    return threshold + 1 # Cap to avoid integer overflow
        return res
